### ISSUE TITLE
Resilver restarts over and over

### ISSUE THREAD

**Issue by lloesche** [zfsonlinux/zfs#4873]
I'm having an issue on my pool that might or might not be ZFS related. I have a 4 HDD (WD Reds 3TB) RaidZ pool where lately during a scrub one of the disks reported one or two CKSUM errors. To make sure it's the disk and not the NAS, cable or controller itself I cycled the drive slots. Sure enough during the next scrub the same HDD in a different slot reported CKSUM errors again. I checked the kernel log and it reported `failed command: WRITE FPDMA QUEUED`. A short and a long smart selftest came back without any errors on all four drives.

Out of ideas I still ordered a new drive. Yesterday I installed the new drive (a 6TB WD Red since they didn't have the 3TB anymore) in the NAS and ran an `zpool replace`. It started to resilver but at around 80% it aborted and began new at 0%.

Again I checked the kernel log and it contained the same error messages it did with the 3TB drive. Again I cycled the drive into a different slot and sure enough after about 11 hours of resilver the drive reported errors in the kernel log again and the resilver restarted.

```
Jul 23 10:44:10 storboy kernel: ata1.00: exception Emask 0x10 SAct 0x8000 SErr 0x4c0001 action 0x6 frozen
Jul 23 10:44:10 storboy kernel: ata1.00: irq_stat 0x08000000, interface fatal error
Jul 23 10:44:10 storboy kernel: ata1: SError: { RecovData CommWake 10B8B Handshk }
Jul 23 10:44:10 storboy kernel: ata1.00: failed command: WRITE FPDMA QUEUED
Jul 23 10:44:10 storboy kernel: ata1.00: cmd 61/08:78:f0:af:a0/00:00:ba:02:00/40 tag 15 ncq 4096 out
                                         res 40/00:70:f0:ad:a0/00:00:ba:02:00/40 Emask 0x10 (ATA bus error)
Jul 23 10:44:10 storboy kernel: ata1.00: status: { DRDY }
Jul 23 10:44:10 storboy kernel: ata1: hard resetting link
Jul 23 10:44:10 storboy kernel: ata1: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
Jul 23 10:44:10 storboy kernel: ata1.00: configured for UDMA/133
Jul 23 10:44:10 storboy kernel: sd 0:0:0:0: [sda] tag#15 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
Jul 23 10:44:10 storboy kernel: sd 0:0:0:0: [sda] tag#15 Sense Key : Illegal Request [current] [descriptor]
Jul 23 10:44:10 storboy kernel: sd 0:0:0:0: [sda] tag#15 Add. Sense: Unaligned write command
Jul 23 10:44:10 storboy kernel: sd 0:0:0:0: [sda] tag#15 CDB: Write(16) 8a 00 00 00 00 02 ba a0 af f0 00 00 00 08 00 00
Jul 23 10:44:10 storboy kernel: blk_update_request: I/O error, dev sda, sector 11721027568
Jul 23 10:44:10 storboy kernel: ata1: EH complete
Jul 23 11:11:02 storboy kernel: ata1.00: exception Emask 0x0 SAct 0x11ff800 SErr 0x4d0001 action 0x6 frozen
Jul 23 11:11:02 storboy kernel: ata1: SError: { RecovData PHYRdyChg CommWake 10B8B Handshk }
Jul 23 11:11:02 storboy kernel: ata1.00: failed command: WRITE FPDMA QUEUED
Jul 23 11:11:02 storboy kernel: ata1.00: cmd 61/50:58:08:22:d6/00:00:09:00:00/40 tag 11 ncq 40960 out
                                         res 40/00:01:00:00:00/00:00:00:00:00/00 Emask 0x4 (timeout)
Jul 23 11:11:02 storboy kernel: ata1.00: status: { DRDY }
Jul 23 11:11:02 storboy kernel: ata1.00: failed command: WRITE FPDMA QUEUED
```

```
[root@blackhole ~]# zpool status
  pool: data
 state: DEGRADED
status: One or more devices is currently being resilvered.  The pool will
    continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Sat Jul 23 18:12:10 2016
    1.09T scanned out of 9.95T at 248M/s, 10h24m to go
    266G resilvered, 10.94% done
config:

    NAME                                            STATE     READ WRITE CKSUM
    data                                            DEGRADED     0     0     0
      raidz1-0                                      DEGRADED     0     0     0
        ata-WDC_WD30EFRX-68EUZN0_WD-WMC4N228XXXX    ONLINE       0     0     0
        ata-WDC_WD30EFRX-68EUZN0_WD-WMC4N235XXXX    ONLINE       0     0     0
        ata-WDC_WD30EFRX-68EUZN0_WD-WMC4N229XXXX    ONLINE       0     0     0
        replacing-3                                 DEGRADED     0     0     0
          13529822044981682645                      UNAVAIL      0     0     0  was /dev/disk/by-id/ata-WDC_WD30EFRX-68EUZN0_WD-WMC4N230XXXX-part1
          ata-WDC_WD60EFRX-68L0BN1_WD-WX31D15FXXXX  ONLINE       0     0     0  (resilvering)

errors: No known data errors
```

I also tried turning off all power management and power saving options just to make sure there's nothing funny going on there. But it didn't make a difference, the drive would resilver to around 80% and then restart, throwing the errors above in the kernel log.

What I did now is, I stopped the resilver, removed the new 6TB drive from the zpool and created an ext4 FS on it. I then rsynced all the data from the degraded pool over to the 6TB drive which finished after 8h without any data errors or further kernel errors in the log.

Any ideas or pointers what might be going on here? It seems that I'm unable to finish a resilver in this pool for this drive even after cycling the drive to two different slots and swapping out the drive for a new one. If need be I can destroy and recreate it but I'd like to find out what's causing these issues.

Additional data:

```
[root@blackhole ~]# lspci -kk
00:00.0 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) Processor Root Complex
    Subsystem: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) Processor Root Complex
00:00.2 IOMMU: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) I/O Memory Management Unit
    Subsystem: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) I/O Memory Management Unit
00:01.0 VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Trinity [Radeon HD 7660D]
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel modules: radeon
00:01.1 Audio device: Advanced Micro Devices, Inc. [AMD/ATI] Trinity HDMI Audio Controller
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: snd_hda_intel
    Kernel modules: snd_hda_intel
00:04.0 PCI bridge: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) Processor Root Port
    Kernel driver in use: pcieport
    Kernel modules: shpchp
00:10.0 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB XHCI Controller (rev 03)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: xhci_hcd
00:10.1 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB XHCI Controller (rev 03)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: xhci_hcd
00:11.0 SATA controller: Advanced Micro Devices, Inc. [AMD] FCH SATA Controller [AHCI mode] (rev 40)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: ahci
00:12.0 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB OHCI Controller (rev 11)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: ohci-pci
00:12.2 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB EHCI Controller (rev 11)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: ehci-pci
00:13.0 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB OHCI Controller (rev 11)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: ohci-pci
00:13.2 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB EHCI Controller (rev 11)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: ehci-pci
00:14.0 SMBus: Advanced Micro Devices, Inc. [AMD] FCH SMBus Controller (rev 14)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: piix4_smbus
    Kernel modules: i2c_piix4, sp5100_tco
00:14.2 Audio device: Advanced Micro Devices, Inc. [AMD] FCH Azalia Controller (rev 01)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device d792
    Kernel driver in use: snd_hda_intel
    Kernel modules: snd_hda_intel
00:14.3 ISA bridge: Advanced Micro Devices, Inc. [AMD] FCH LPC Bridge (rev 11)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
00:14.4 PCI bridge: Advanced Micro Devices, Inc. [AMD] FCH PCI Bridge (rev 40)
00:14.5 USB controller: Advanced Micro Devices, Inc. [AMD] FCH USB OHCI Controller (rev 11)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: ohci-pci
00:18.0 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) Processor Function 0
00:18.1 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) Processor Function 1
00:18.2 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) Processor Function 2
00:18.3 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) Processor Function 3
    Kernel driver in use: k10temp
    Kernel modules: k10temp
00:18.4 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) Processor Function 4
00:18.5 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 15h (Models 10h-1fh) Processor Function 5
01:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller (rev 06)
    Subsystem: Micro-Star International Co., Ltd. [MSI] Device 7792
    Kernel driver in use: r8169
    Kernel modules: r8169

[root@blackhole ~]# uname -a
Linux blackhole 4.6.4-201.fc23.x86_64 #1 SMP Tue Jul 12 11:43:59 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

[root@blackhole ~]# cat /etc/redhat-release
Fedora release 23 (Twenty Three)

[root@blackhole ~]# rpm -qa zfs spl
spl-0.6.5.7-1.fc23.x86_64
zfs-0.6.5.7-1.fc23.x86_64

```

**Comment by dasjoe** [comment]
Hi, what you're seeing here looks like disk errors. I'd say this is not a bug in ZFS, so I'd rather take the discussion to #zfsonlinux on freenode or the mailing list.

Check out https://gist.github.com/dasjoe/a7252e3f737ec36de36f for a quick way to collect (hopefully relevant) SMART stats.

**Comment by kernelOfTruth** [comment]
>The logs to me all look like it's a kernel or controller issue - or if I hadn't replaced the HDD for a new one a disk issue. But the thing is it only happens with ZFS and always on the same pool device irregardless of where the disk is located in the system or what kind of disk I'm using.

well, ZFS is more pedantic towards hardware errors :)

Please take the SATA cables into consideration whether they are okay, too.

Those errors look like:
harddrive failing/failure, cable (connection) issues, controller <-- cable --> harddrive issues,

all of these I've already experienced before afaik, so be wary of hardware failure (mainboard, sata controller, cable, harddrive, etc.)

**Comment by lloesche** [comment]
That's exactly what I was thinking, but then it doesn't make sense that the error follows the pool device around in the NAS. Like, originally the drive was in bay 1. I swapped it to bay 2 and the error followed. I replaced the drive with a brand new one, same error. I cycled it to bay 3 again the error followed. I formatted the new 6TB drive with ext4 and copied all the data (took like 15h) no errors whatsoever. Started the resilver again, error reappeared. Only when I rebooted the system once per hour I had zero errors in the log over a period of 20 hours and resilver finished. It's the damnedest thing.

**Comment by kernelOfTruth** [comment]
Hm, in that case it's the "high i/o pressure - i/o errors are being spit" situation,

factors that caused this for me were:

failing harddrive, faulty/moody cable and controller and driver (both SATA or USB 3.0/xhci), filesystem bugs (other filesystem than ZFS in my case)

potential additional factors that I could think of and partially already experienced:
i/o scheduler, encryption (luks, dm-crypt), kernel version with some inherent bug which didn't get backported, some weird processor or platform related issue (AMD or Intel), NCQ (set to 1 queue_depth to disable or force-disable it via kernel append), nr_requests, ...

**Comment by danmac-uk** [comment]
I'm seeing this same issue on a raidz1 with 3 Hitachi 2TB drives on 4.4 kernel (Proxmox) on an ICH10R. Drives pass badblocks -wsv fully and are SMART clear but are otherwise not known good, cables have been swapped. I have seen a small (tens) number of UDMA CRC errors on all drives but not enough for a collision :P Drives are 512-byte native but the array is currently set to ashift=12, I plan to test with ashift=9. RAM passes memtest and is known good for several years but is not ECC. *hangs head in shame*

100% reproduceable with current config but takes a couple of hours. It seems like ZFS is writing to the disks, getting the SATA errors, leaving the data corrupted, then subsequently fixing it with scrub. I have yet to trough through the logs and establish the full timeline etc. I have seen these errors occur on two of the three drives in the array ie. the first two of sdb, sdc, sdd. I also plan to swap sdb / sdd around and re-test, if it's a controller port issue. I doubt it as the sdb port (2/6) is known good, but you never know.

Happy to test any theories - at the moment the machine has nothing on it that tar & netcat can't put back, and while I would like to get it set-up I'm not in a massive rush :) Please let me know if there's anything I can help with to track this one down.

Cheers,

Dan.

**Comment by danmac-uk** [comment]
The following did not help: cables changed, ports changed, power supply changed, ashift=9 instead of 12, NCQ forced off.

What appears to have fixed it is forcing the SATA speed to 1.5Gbps. (kernel boot param `libata.force=1.5`)

I'm proceeding under the assumption the ICH10R SATA controller doesn't entirely like these Hitachi drives, and somewhere within the various layers of abstraction something is failing to retry the write when that's clearly the right thing to do. Data ends up corrupted on disk as a result, and ZFS only becomes aware of it on the next read.

In my humble opinion this could constitute at least two bugs - write errors ignored by ZFS and transient SATA link issues not handled correctly. I would say the 1st is a ZFS bug and the 2nd is perhaps a libata bug although I'm unsure where in the stack the buck stops?

Test script I've been using here:
https://gist.github.com/danmac-uk/a71330713fd65c99c6d0baa27567c130

**Comment by richardelling** [comment]
ZFS relies on the underlying driver to propagate an error. In other words, if sd (or whatever driver is below) does not report an error to ZFS, ZFS expect things are ok.

What we see in cases where SATA fabrics corrupt data is that the error is detected later during a scrub. Note that the read can also be corrupted, so when the underlying driver delivers a read without error to ZFS, then ZFS considers that read to be good, and thus assumes any data corruption is permanent at the media. If ZFS repairs this data, then there is a non-zero probability that the repair doesn't affect the outcome. If ZFS cannot repair the data, subsequent reads might be fine. Such problems with SATA or SAS fabrics can be maddening to troubleshoot, but there really isn't anything ZFS can do if the underlying drivers and layers propagate bad data without also indicating failure.

**Comment by acertain** [comment]
If anybody is interested, I'm seeing the same things, also with some WD drives. I'm more paranoid than the OP, so when I built my array, I ordered two WD drives and two HGST drives (I realize that WD owns HGST, but at least at the time they were different designs, firmware, fab lines, etc.). Two days ago, I was down in the computer closet and heard the sound of a hard-drive resetting. I looked in dmesg and saw errors from both WD drives. They both would record errors at the same time and then carry on for a bit.

When I ran the zfs scrub, it found a bunch of errors and repaired them, but reported no disk errors. I then looked at the disk SMART data and found no errors, no pending remapped sectors, etc. Given that I'm also running ZFS on Linux, these two drives are on separate controllers (each controller handles one WD and one HGST drive), there are no SMART errors reported on any disk, and the errors for the two WD drives always happen at the same time (and no errors for the HGST drives), it seems likely that there's something weird in the ZFS/WD/SATA integration.

Here's one example sequence from dmesg. I also saw some with "Unaligned write command:"
[May11 18:45] ata1: lost interrupt (Status 0x50)
[  +0.000002] ata3: lost interrupt (Status 0x50)
[  +0.000017] ata3.00: exception Emask 0x10 SAct 0x0 SErr 0x40d0002 action 0xe frozen
[  +0.000003] ata1.00: exception Emask 0x50 SAct 0x0 SErr 0x40d0802 action 0xe frozen
[  +0.000005] ata1.00: SError: { RecovComm HostInt PHYRdyChg CommWake 10B8B DevExch }
[  +0.000003] ata1.00: failed command: FLUSH CACHE EXT
[  +0.000007] ata1.00: cmd ea/00:00:00:00:00/00:00:00:00:00/a0 tag 0
                       res 40/00:01:00:00:00/00:00:00:00:00/40 Emask 0x54 (ATA bus error)
[  +0.000002] ata1.00: status: { DRDY }
[  +0.000012] ata1.00: hard resetting link
[  +0.011605] ata3: SError: { RecovComm PHYRdyChg CommWake 10B8B DevExch }
[  +0.001670] ata3.00: failed command: FLUSH CACHE EXT
[  +0.001677] ata3.00: cmd ea/00:00:00:00:00/00:00:00:00:00/a0 tag 0
                       res 40/00:01:00:00:00/00:00:00:00:00/40 Emask 0x14 (ATA bus error)
[  +0.003385] ata3.00: status: { DRDY }
[  +0.001612] ata3: hard resetting link
[  +0.704057] ata1.01: hard resetting link
[  +0.176042] ata3: SATA link up 3.0 Gbps (SStatus 123 SControl 300)
[  +0.024373] ata3.00: configured for UDMA/133
[  +0.000005] ata3.00: retrying FLUSH 0xea Emask 0x14
[  +0.000190] ata3: EH complete
[  +0.827435] ata1.01: failed to resume link (SControl 0)
[  +0.156079] ata1.00: SATA link up 3.0 Gbps (SStatus 123 SControl 300)
[  +0.000015] ata1.01: SATA link down (SStatus 0 SControl 0)
[  +0.024442] ata1.00: configured for UDMA/133
[  +0.000005] ata1.00: retrying FLUSH 0xea Emask 0x54
[  +0.000209] ata1: EH complete

I just ran a second scrub and this time zfs is reporting a http://zfsonlinux.org/msg/ZFS-8000-9P/ error on one of the HGST drives. There's nothing in dmsg or SMART, but it's making me nervous. I have two replacement drives here, and I'll probably drop them this weekend just to give me more piece of mind. But if there's anything people would like for me to try to troubleshoot, I'd be happy to do it. I'm running with an Intel SATA controller ("6 Series/C200 Series Chipset Family 2 port SATA IDE Controller" as reported by lshw) and my drives are WDC WD40EZRZ-00W. My kernel is 4.4.0-77-generic and I've got version 0.6.5.6-0ubuntu16 of the Linux ZFS packages.

Andrew

**Comment by danmac-uk** [comment]
As I detailed in my previous comment the only way to make ZFS on Linux not trash the occasional write on my system (3x HGST drives in a RAIDZ) is to limit SATA speed to 1.5Gbps. I haven't had a problem since I did that.

**Comment by ccoager** [comment]
Wanted to post that I'm having this same issue with a brand new disk rebuilding a raid mirror.  This is a Samsung SSD 860 PRO 256GB with 55 power on hours.  Did a long SMART test, overall health test is pass.  I tried swapping the SATA cable and the port on the motherboard with no improvement.

part of dmesg:
> [48201.699494] sd 6:0:0:0: [sdq] tag#15 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[48201.699500] sd 6:0:0:0: [sdq] tag#15 Sense Key : Illegal Request [current]
[48201.699506] sd 6:0:0:0: [sdq] tag#15 Add. Sense: Unaligned write command
[48201.699511] sd 6:0:0:0: [sdq] tag#15 CDB: Write(10) 2a 00 12 20 c3 80 00 00 40 00
[48201.699514] print_req_error: I/O error, dev sdq, sector 304137088
[48201.704103] ata6: EH complete
[48201.704411] ata6.00: Enabling discard_zeroes_data
[48202.866980] ata6.00: exception Emask 0x10 SAct 0x7000007f SErr 0x0 action 0x6 frozen
[48202.871281] ata6.00: irq_stat 0x08000000, interface fatal error
[48202.875656] ata6.00: failed command: WRITE FPDMA QUEUED
[48202.879908] ata6.00: cmd 61/10:00:e0:f5:24/00:00:12:00:00/40 tag 0 ncq dma 8192 out
                        res 40/00:e0:70:f5:24/00:00:12:00:00/40 Emask 0x10 (ATA bus error)
[48202.888464] ata6.00: status: { DRDY }

zpool status:
>   pool: rpool
 state: DEGRADED
status: One or more devices are faulted in response to persistent errors.
        Sufficient replicas exist for the pool to continue functioning in a
        degraded state.
action: Replace the faulted device, or use 'zpool clear' to mark the device
        repaired.
  scan: resilvered 136M in 0h15m with 0 errors on Mon Dec 24 21:27:26 2018
config:

        NAME                                                     STATE     READ WRITE CKSUM
        rpool                                                    DEGRADED     0     0     0
          mirror-0                                               DEGRADED     0     0     0
            ata-Samsung_SSD_860_PRO_256GB_S418NF0K911700H-part3  FAULTED      0     0     0  too many errors
            ata-Samsung_SSD_850_PRO_256GB_S39KNX0J709139J-part3  ONLINE       0     0     0

errors: No known data errors

**Comment by ccoager** [comment]
Tried swapping the power cables, no luck.  Tried using libata.force=6:1.5G, no luck.

**Comment by danmac-uk** [comment]
Merry Christmas :) Are you sure the SSD is negotiating 1.5Gbps SATA speed as the excerpt of your log doesn't show that? Try "libata.force=1.5"? (this should limit all SATA ports to 1.5Gbps) Maybe swap the SATA port you're connecting it to? Good luck :)

**Comment by ccoager** [comment]
@danmac-uk The drive in question is currently on ata6 which I have limited to 1.5G.  I tried swapping the port already, didn't make a difference.

> [32448.092503] ata6.00: exception Emask 0x10 SAct 0xffc000 SErr 0x0 action 0x6 frozen
[32448.092617] ata6.00: irq_stat 0x08000000, interface fatal error
[32448.092702] ata6.00: failed command: WRITE FPDMA QUEUED
[32448.092782] ata6.00: cmd 61/10:70:b0:9a:3a/00:00:11:00:00/40 tag 14 ncq dma 8192 out
                        res 40/00:70:b0:9a:3a/00:00:11:00:00/40 Emask 0x10 (ATA bus error)
[32448.092991] ata6.00: status: { DRDY }
[32448.093045] ata6.00: failed command: WRITE FPDMA QUEUED
[32448.093124] ata6.00: cmd 61/10:78:c0:9a:3a/00:00:11:00:00/40 tag 15 ncq dma 8192 out
                        res 40/00:70:b0:9a:3a/00:00:11:00:00/40 Emask 0x10 (ATA bus error)
[32448.093333] ata6.00: status: { DRDY }
[32448.093386] ata6.00: failed command: WRITE FPDMA QUEUED
[32448.093465] ata6.00: cmd 61/50:80:d0:9a:3a/00:00:11:00:00/40 tag 16 ncq dma 40960 out
                        res 40/00:70:b0:9a:3a/00:00:11:00:00/40 Emask 0x10 (ATA bus error)
[32448.093673] ata6.00: status: { DRDY }
[32448.093726] ata6.00: failed command: WRITE FPDMA QUEUED
[32448.093805] ata6.00: cmd 61/70:88:20:9b:3a/00:00:11:00:00/40 tag 17 ncq dma 57344 out
                        res 40/00:70:b0:9a:3a/00:00:11:00:00/40 Emask 0x10 (ATA bus error)
[32448.094013] ata6.00: status: { DRDY }
[32448.094066] ata6.00: failed command: WRITE FPDMA QUEUED
[32448.094145] ata6.00: cmd 61/10:90:90:9b:3a/00:00:11:00:00/40 tag 18 ncq dma 8192 out
                        res 40/00:70:b0:9a:3a/00:00:11:00:00/40 Emask 0x10 (ATA bus error)
[32448.094353] ata6.00: status: { DRDY }
[32448.094406] ata6.00: failed command: WRITE FPDMA QUEUED
[32448.094485] ata6.00: cmd 61/10:98:a0:9b:3a/00:00:11:00:00/40 tag 19 ncq dma 8192 out
                        res 40/00:70:b0:9a:3a/00:00:11:00:00/40 Emask 0x10 (ATA bus error)
[32448.094692] ata6.00: status: { DRDY }
[32448.094746] ata6.00: failed command: WRITE FPDMA QUEUED
[32448.094842] ata6.00: cmd 61/50:a0:b0:9b:3a/00:00:11:00:00/40 tag 20 ncq dma 40960 out
                        res 40/00:70:b0:9a:3a/00:00:11:00:00/40 Emask 0x10 (ATA bus error)
[32448.095054] ata6.00: status: { DRDY }
[32448.095108] ata6.00: failed command: WRITE FPDMA QUEUED
[32448.095187] ata6.00: cmd 61/30:a8:00:9c:3a/00:00:11:00:00/40 tag 21 ncq dma 24576 out
                        res 40/00:70:b0:9a:3a/00:00:11:00:00/40 Emask 0x10 (ATA bus error)
[32448.095395] ata6.00: status: { DRDY }
[32448.095448] ata6.00: failed command: WRITE FPDMA QUEUED
[32448.095527] ata6.00: cmd 61/10:b0:30:9c:3a/00:00:11:00:00/40 tag 22 ncq dma 8192 out
                        res 40/00:70:b0:9a:3a/00:00:11:00:00/40 Emask 0x10 (ATA bus error)
[32448.095735] ata6.00: status: { DRDY }
[32448.095790] ata6.00: failed command: WRITE FPDMA QUEUED
[32448.095869] ata6.00: cmd 61/40:b8:40:9c:3a/00:00:11:00:00/40 tag 23 ncq dma 32768 out
                        res 40/00:70:b0:9a:3a/00:00:11:00:00/40 Emask 0x10 (ATA bus error)
[32448.096077] ata6.00: status: { DRDY }
[32448.096132] ata6: hard resetting link
[32448.580540] ata6: SATA link up 1.5 Gbps (SStatus 113 SControl 310)
[32448.582575] ata6.00: supports DRM functions and may not be fully accessible
[32448.586226] ata6.00: supports DRM functions and may not be fully accessible
[32448.589228] ata6.00: configured for UDMA/133
[32448.589252] scsi_io_completion: 10 callbacks suppressed
[32448.589260] sd 6:0:0:0: [sdq] tag#14 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[32448.589266] sd 6:0:0:0: [sdq] tag#14 Sense Key : Illegal Request [current]
[32448.589271] sd 6:0:0:0: [sdq] tag#14 Add. Sense: Unaligned write command
[32448.589277] sd 6:0:0:0: [sdq] tag#14 CDB: Write(10) 2a 00 11 3a 9a b0 00 00 10 00
[32448.589279] print_req_error: 10 callbacks suppressed
[32448.589281] print_req_error: I/O error, dev sdq, sector 289053360
[32448.589392] sd 6:0:0:0: [sdq] tag#15 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[32448.589394] sd 6:0:0:0: [sdq] tag#15 Sense Key : Illegal Request [current]
[32448.589396] sd 6:0:0:0: [sdq] tag#15 Add. Sense: Unaligned write command
[32448.589398] sd 6:0:0:0: [sdq] tag#15 CDB: Write(10) 2a 00 11 3a 9a c0 00 00 10 00
[32448.589399] print_req_error: I/O error, dev sdq, sector 289053376
[32448.592796] sd 6:0:0:0: [sdq] tag#16 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[32448.592798] sd 6:0:0:0: [sdq] tag#16 Sense Key : Illegal Request [current]
[32448.592801] sd 6:0:0:0: [sdq] tag#16 Add. Sense: Unaligned write command
[32448.592803] sd 6:0:0:0: [sdq] tag#16 CDB: Write(10) 2a 00 11 3a 9a d0 00 00 50 00
[32448.592804] print_req_error: I/O error, dev sdq, sector 289053392
[32448.596202] sd 6:0:0:0: [sdq] tag#17 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[32448.596204] sd 6:0:0:0: [sdq] tag#17 Sense Key : Illegal Request [current]
[32448.596206] sd 6:0:0:0: [sdq] tag#17 Add. Sense: Unaligned write command
[32448.596208] sd 6:0:0:0: [sdq] tag#17 CDB: Write(10) 2a 00 11 3a 9b 20 00 00 70 00
[32448.596210] print_req_error: I/O error, dev sdq, sector 289053472
[32448.599695] sd 6:0:0:0: [sdq] tag#18 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[32448.599701] sd 6:0:0:0: [sdq] tag#18 Sense Key : Illegal Request [current]
[32448.599706] sd 6:0:0:0: [sdq] tag#18 Add. Sense: Unaligned write command
[32448.599713] sd 6:0:0:0: [sdq] tag#18 CDB: Write(10) 2a 00 11 3a 9b 90 00 00 10 00
[32448.599717] print_req_error: I/O error, dev sdq, sector 289053584
[32448.603026] sd 6:0:0:0: [sdq] tag#19 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[32448.603028] sd 6:0:0:0: [sdq] tag#19 Sense Key : Illegal Request [current]
[32448.603030] sd 6:0:0:0: [sdq] tag#19 Add. Sense: Unaligned write command
[32448.603032] sd 6:0:0:0: [sdq] tag#19 CDB: Write(10) 2a 00 11 3a 9b a0 00 00 10 00
[32448.603033] print_req_error: I/O error, dev sdq, sector 289053600
[32448.606233] sd 6:0:0:0: [sdq] tag#20 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[32448.606236] sd 6:0:0:0: [sdq] tag#20 Sense Key : Illegal Request [current]
[32448.606238] sd 6:0:0:0: [sdq] tag#20 Add. Sense: Unaligned write command
[32448.606240] sd 6:0:0:0: [sdq] tag#20 CDB: Write(10) 2a 00 11 3a 9b b0 00 00 50 00
[32448.606242] print_req_error: I/O error, dev sdq, sector 289053616
[32448.609488] sd 6:0:0:0: [sdq] tag#21 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[32448.609491] sd 6:0:0:0: [sdq] tag#21 Sense Key : Illegal Request [current]
[32448.609492] sd 6:0:0:0: [sdq] tag#21 Add. Sense: Unaligned write command
[32448.609495] sd 6:0:0:0: [sdq] tag#21 CDB: Write(10) 2a 00 11 3a 9c 00 00 00 30 00
[32448.609496] print_req_error: I/O error, dev sdq, sector 289053696
[32448.612716] sd 6:0:0:0: [sdq] tag#22 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[32448.612719] sd 6:0:0:0: [sdq] tag#22 Sense Key : Illegal Request [current]
[32448.612721] sd 6:0:0:0: [sdq] tag#22 Add. Sense: Unaligned write command
[32448.612724] sd 6:0:0:0: [sdq] tag#22 CDB: Write(10) 2a 00 11 3a 9c 30 00 00 10 00
[32448.612725] print_req_error: I/O error, dev sdq, sector 289053744
[32448.615927] sd 6:0:0:0: [sdq] tag#23 FAILED Result: hostbyte=DID_OK driverbyte=DRIVER_SENSE
[32448.615929] sd 6:0:0:0: [sdq] tag#23 Sense Key : Illegal Request [current]
[32448.615931] sd 6:0:0:0: [sdq] tag#23 Add. Sense: Unaligned write command
[32448.615933] sd 6:0:0:0: [sdq] tag#23 CDB: Write(10) 2a 00 11 3a 9c 40 00 00 40 00
[32448.615934] print_req_error: I/O error, dev sdq, sector 289053760
[32448.619141] ata6: EH complete
[32448.619536] ata6.00: Enabling discard_zeroes_data

**Comment by sitsofe** [comment]
@ccoager  I've just hit the same issue during a ZFS resilver with a new Samsung 860 EVO too. Do you happen to be using an AMD motherboard?

**Comment by sitsofe** [comment]
@ccoager Me too (inside an HP microserver). There's a curious rumour going round about issues with AMD motherboards and Samsung SSDs e.g. https://eu.community.samsung.com/t5/Cameras-IT-Everything-Else/860-EVO-250GB-causing-freezes-on-AMD-system/td-p/575813 . There are grumbles about older Samsung SSDs and AMD chipsets too - http://www.tomshardware.co.uk/answers/id-2989221/amd-chipset-issues-sam-850-pro-ssd.html . It's all a bit hearsay but some seem to have solved their issue but going with a different SSD vendor or much much older SSD versions...

**Comment by sitsofe** [comment]
@ccoager My OS is Ubuntu 18.04 with a 4.15.0-43-generic kernel (but even a 4.18.0-13-generic had the same issue). I went on to disable NCQ for just the port that the Samsung SSD was on (in my case by adding `libata.force=2.00:noncq` to the boot command line) and that stopped the errors on my particular system. Without that, even a simple read fio job like:

```
fio --name=test --readonly --rw=randread --filename /dev/sdb --bs=32k --ioengine=libaio --iodepth=32 --direct=1 --runtime=10m --time_based=1
```

would trigger `failed command: WRITE FPDMA QUEUED` errors and bus resets.

There's a definite speed hit to disabling NCQ and the smaller the block sizes being used the bigger the hit (compared to the top speed) but not triggering errors is kind of key. I tried swapping the SSD to another caddy and another bay but the issue persists. Some out on the internet say this could be down to a failing power supply but it seems a bit strange that a Crucial MX500 SSD in the same system has never had an issue...
